src/_P007_PCF8591.ino

```diff
@@ -81,39 +148,50 @@ boolean Plugin_007(uint8_t function, struct EventStruct *event, String& string)
       break;
     }
 
-    case PLUGIN_I2C_HAS_ADDRESS:
-    {
-      const int i2cAddressValues[] = { 0x48, 0x49, 0x4a, 0x4b, 0x4c, 0x4d, 0x4e, 0x4f };
-      success = intArrayContains(8, i2cAddressValues, event->Par1);
-      break;
-    }
-
     case PLUGIN_READ:
     {
       uint8_t unit    = (CONFIG_PORT - 1) / 4;
       uint8_t port    = CONFIG_PORT - (unit * 4);
       uint8_t address = 0x48 + unit;
 
-      // get the current pin value
-      Wire.beginTransmission(address);
-      Wire.write(port - 1);
-      Wire.endTransmission();
-
-      Wire.requestFrom(address, (uint8_t)0x2);
-
-      if (Wire.available())
-      {
-        Wire.read();                                       // Read older value first (stored in chip)
-        UserVar[event->BaseVarIndex] = (float)Wire.read(); // now read actual value and store into Nodo var
-
-        if (loglevelActiveFor(LOG_LEVEL_INFO)) {
-          String log;
-          log.reserve(30);
-          log = F("PCF  : Analog value: ");
-          log += formatUserVarNoCheck(event->TaskIndex, 0);
-          addLog(LOG_LEVEL_INFO, log);
+      uint8_t var = 0;
+
+      for (; var < P007_NR_OUTPUT_VALUES; ++port, ++var) {
+        if (port <= 4) { // Only read available ports, hardwired limited to 4
+          Wire.beginTransmission(address);
+
+          // get the current pin value
+          Wire.write(port - 1);
+          Wire.endTransmission();
+
+          Wire.requestFrom(address, (uint8_t)0x2);
+
+          if (Wire.available())
+          {
+            Wire.read();                                             // Read older value first (stored in chip)
+            UserVar[event->BaseVarIndex + var] = (float)Wire.read(); // now read actual value and store into Value var
```

Comment:
How well does this work?
Is there always an old value and a new value?
Is the new value immediately available after you request it?
Is it possible due to timing issues that you're reading a value from another port?

Reply:
The first `read()` is reading the status/result of a previous read action, so not really interesting. After that, the read ADC converted data can be read (8 bits only). The 'polling-limit' is only limited by the I2C bus speed, according to the documentation.
How accurate that documentation and the actual chip really match I can't predict... 🤔